contracts/upgradeable_contracts/arbitrary_message/MessageProcessor.sol

```diff
@@ -0,0 +1,79 @@
+pragma solidity 0.4.24;
+
+import "./BalanceHandler.sol";
+import "../../libraries/ArbitraryMessage.sol";
+
+
+contract MessageProcessor is BalanceHandler {
+
+    uint256 internal constant PASS_MESSAGE_GAS = 100000;
+
+    function processMessage(bytes _data, bool applyDataOffset) internal {
+        address sender;
+        address executor;
+        uint256 gasLimit;
+        bytes1 dataType;
+        uint256 gasPrice;
+        bytes32 txHash;
+        bytes memory data;
+        (sender, executor, txHash, gasLimit, dataType, gasPrice, data) = ArbitraryMessage.unpackData(_data, applyDataOffset);
+
+        require(!messageProcessed(txHash));
+        setMessageProcessed(txHash, true);
+
+        if (dataType == 0x00) {
+            require(isMessageProcessorSubsidizedMode());
+            _passMessage(sender, executor, data, gasLimit);
+        } else if (dataType == 0x01) {
+            require(!isMessageProcessorSubsidizedMode());
+            require(gasPrice == tx.gasprice);
+            _defrayAndPassMessage(sender, executor, data, gasLimit);
+        } else if (dataType == 0x02) {
+            require(!isMessageProcessorSubsidizedMode());
+            _defrayAndPassMessage(sender, executor, data, gasLimit);
+        } else {
+            revert();
+        }
+
+        emitEventOnMessageProcessed(sender, executor, txHash);
+    }
+
+    function _defrayAndPassMessage(address _sender, address _contract, bytes _data, uint256 _gas) internal {
+        uint256 fee = PASS_MESSAGE_GAS.add(_gas).mul(tx.gasprice);
+        require(balanceOf(_sender) >= fee);
+        require(address(this).balance >= fee);
+
+        setBalanceOf(_sender, balanceOf(_sender).sub(fee));
+
+        _passMessage(_sender, _contract, _data, _gas);
+
+        msg.sender.transfer(fee);
```

Comment:
@akolotov does it makes sense to emit some kind of `transfer` event after sending fee to validator?

Reply:
No, it is not necessary since this transfer assumes to just cover the message sender expenses to send this transaction